src/libexpr/primops/regex.cc

```diff
@@ -0,0 +1,360 @@
+{
+    // Cache the string in the match data for match[] to work.
+    match.str = str;
+    // compile if we're using this regex more than once
+    if (this->usage++) this->compile();
+
+    int rc = pcre2_match(code, (const unsigned char*)str.data(), str.length(),
+        startoffset, options, match.match, NULL);
+
+    assert(rc != 0); // match data too small, shouldn't happen
+
+    if (rc < 0 && rc != PCRE2_ERROR_NOMATCH) {
+        unsigned char err[256];
+        pcre2_get_error_message(rc, err, sizeof(err));
+        throw RegexError("unable to match regex: %1%", err);
+    }
+    return rc;
+};
+
+};
+
+struct RegexCache
+{
+    // TODO use C++20 transparent comparison when available
+    std::unordered_map<std::string_view, PCRE::Regex> cache;
+    std::list<std::string> keys;
+
+    PCRE::Regex& get(std::string_view re)
+    {
+        auto it = cache.find(re);
+        if (it != cache.end())
+            return it->second;
+        keys.emplace_back(re);
+        return cache.emplace(keys.back(), keys.back()).first->second;
+    }
+};
+
+std::shared_ptr<RegexCache> makeRegexCache()
+{
+    return std::make_shared<RegexCache>();
+}
+
+size_t regexCacheSize(std::shared_ptr<RegexCache> cache)
+{
+    return cache->keys.size();
+}
+
+static void extract_matches(EvalState & state, const PCRE::MatchData& match, Value & v)
+{
+    auto nameTable = match.regex().nameTable();
+    if (nameTable.size()) {
+        // try to extract named bindings
+        auto bindings = state.buildBindings(nameTable.size());
+        for (auto i : nameTable) {
+            Value & elem = bindings.alloc(i.first);
+            if (!match[i.second].has_value()) elem.mkNull();
+            else elem.mkString(*match[i.second]);
+        }
+        v.mkAttrs(bindings);
+    } else {
+        // the first match is the whole string
+        const size_t len = match.size() - 1;
+        state.mkList(v, len);
+        for (size_t i = 0; i < len; ++i) {
+            if (!match[i+1].has_value())
+                (v.listElems()[i] = state.allocValue())->mkNull();
+            else
+                (v.listElems()[i] = state.allocValue())->mkString(*match[i + 1]);
+        }
+    }
+}
+
+void prim_match(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto re = state.forceStringNoCtx(*args[0], pos);
+
+    try {
+
+        auto& regex = state.regexCache->get(re);
+
+        PathSet context;
+        const auto str = state.forceString(*args[1], context, pos);
+
+        PCRE::MatchData match(regex);
+        int rc = regex.match(str, match, 0, PCRE2_ANCHORED | PCRE2_ENDANCHORED);
+        if (rc == PCRE2_ERROR_NOMATCH) {
+            v.mkNull();
+            return;
+        }
+        extract_matches(state, match, v);
+
+    } catch (RegexError & e) {
+        state.debugThrowLastTrace(EvalError({
+            .msg = hintfmt("error while evaluating regex '%s': ", re, e.what()),
+            .errPos = state.positions[pos]
+        }));
```

Comment:
Actually the above doesn't look so great, maybe rather something like

```suggestion
        e.addTrace(state.positions[pos], "while evaluating regex '%s'", re);
        state.debugThrowLastTrace(e);
```